### PR TITLE
Estimate Remaining Time (graphical display)

### PR DESCRIPTION
**Pull request by LinFor**
This is currently in draft so I can do more testing on real hardware.

### Description

1. Added optional time-to-print-completion estimation based on current elapsed print time and current progress estimation. It prints together with elapsed time via blinking, estimated completion time prefixed with "E" letter. (PRINT_PROGRESS_ESTIMATE_TIME_TO_COMPLETION)
2. To increase precision of estimation added ability to calculate progress in permyads (1/100 of percent and 1/10000 of whole). This is done in fixed-point math via uint16 (instead of uint8).
3. Added optional ability to show calculated print progress with up to 2 decimal digits (PRINT_PROGRESS_SHOW_DECIMALS).
4. M73 command now can accept fractional P argument to support above scenarios.
5. Implemented caching of print progress / elapsed time / estimated completion time calculation results to decrease CPU cycles usage. Buffers are prepared during the "first-page" preparation and updates only if progress changed or 1 second has passed.

All of these is done only for DOGM displays, but possible for another ones (but I hasn't somthing another hardware for testing).

### Benefits

Ability to estimate time to print completion.
Ability to more precise estimate print progress for long-running prints.

### Demo

<a href="http://www.youtube.com/watch?feature=player_embedded&v=j9oTQdBIaiE" target="_blank">demonstration</a>

### Related Issues

#12123, #8515, #2181

### CODE REVIEW

File: Marlin/src/gcode/lcd/M73.cpp [hunk 1]
```diff
@@ -39,7 +39,11 @@
  */
 void GcodeSuite::M73() {
   if (parser.seen('P'))
-    ui.set_progress(parser.value_byte());
+    #if HAS_PRINT_PROGRESS_PERMYRIAD
+      ui.set_progress((uint16_t)(parser.value_float() * 100));
+    #else
+      ui.set_progress(parser.value_byte());
+    #endif
```
Comment (thinkyhead) [inline Marlin/src/gcode/lcd/M73.cpp]: ~~G-codes can't change the meaning of their parameters without also exposing this fact to hosts and users. Thus you will have to use a parameter other than 'P' for this new feature.~~

**_Nevermind. I see now._**

File: Marlin/src/lcd/ultralcd.h [hunk 2]
```diff
@@ -292,26 +292,28 @@ class MarlinUI {
...
-          static void set_progress_done() { progress_override = 0x8000 + 10000; }
-          static void progress_reset() { if (progress_override & 0x8000) set_progress_permyriad(0); }
-        #elif HAS_PRINT_PROGRESS_PERCENT
-          static uint8_t progress_override;
-          static void set_progress(const uint8_t progress) { progress_override = _MIN(progress, 100); }
-          static void set_progress_done() { progress_override = 0x80 + 100; }
-          static void progress_reset() { if (progress_override & 0x80) set_progress(0); }
-        #endif
+        static progress_t progress_override;
+        static void set_progress(const uint8_t p) { progress_override = _MIN(p, 100) * PROGRESS_SCALE; }
```
Comment (LinFor) [inline Marlin/src/lcd/ultralcd.h]: I want to have ability to set fractional percents via M73 Pxx.xx, as this done in klipper. So for this reason I included specific set_progress_permyriad() with uint16

Comment (LinFor) [reply Marlin/src/lcd/ultralcd.h]: Now made with single set_progress and PROGRESS_SCALE, as set_progress used only by M73

**Comment by LinFor** [comment]
I have a strange issue with LCD and seems that this PR not related to them.

When I print from SD file **without** M73 commands in it - LCD updates as usual, once per second: https://youtu.be/6oVtPs9nY8I
When I print from SD same file but **with** M73 commands - LCD starts to update very quickly. This is visible in X/Y/Z position counters and temperatures - which code I didn't touched at all: https://youtu.be/ktYBHKg6h7A (sorry for video orientation - i don't know how rotate video on youtube quickly).

Problem is that later LCD freezes - probably due to very high mcu load from lcd calculation and drawing.

I have no idea how this matters.

**Comment by LinFor** [comment]
I made some investigation about "quick LCD updating". It's not relevant to this PR or M73 command at all, instead it occurs after any M117 command. Probably, we must create a bug.
Also, LCD freezes gone after increasing DOGM_SPI_DELAY_US up to 15, so it's problem with my hardware.

File: Marlin/src/lcd/ultralcd.h (hunk 2 above)
Comment (thinkyhead) [reply Marlin/src/lcd/ultralcd.h]: I am preserving that, as can be seen in `M73`. I am also preserving this.

Comment (thinkyhead) [reply Marlin/src/lcd/ultralcd.h]: Ok, sounds good.

**Comment by thinkyhead** [comment]
> Probably, we must create a bug.

The solution is to comment out the `refresh()` at the end of `MarlinUI::finish_status`.

File: Marlin/src/libs/numtostr.h [hunk 3]
```diff
@@ -32,6 +32,11 @@ char* ui8tostr3(const uint8_t i);
 // Convert int8_t to string with 123 format
 char* i8tostr3(const int8_t x);
 
+#if HAS_PRINT_PROGRESS_PERMYRIAD
+  // Convert unsigned 16bit fixed-point float (in hundredth parts of int) to string 100 / 23 / 23.4 / 3.45 format
+  char* ui16fptostr4(const uint16_t xx);
```
Comment (thinkyhead) [inline Marlin/src/libs/numtostr.h]: Note that this naming is misleading. Fixed-point represents fractions but is still in base-2. This is not fixed-point, but simply a 100x scaled integer value.

**Comment by randellhodges** [comment]
I think there is an issue around this line:
https://github.com/MarlinFirmware/Marlin/blob/b20d5a7cc627164877c95859baaa614f8afc7b99/Marlin/src/libs/numtostr.cpp#L206

```
Marlin\src\libs\numtostr.cpp: In function 'char* ftostr4sign(const float&)':
Marlin\src\libs\numtostr.cpp:206:51: error: invalid conversion from 'const char*' to 'char*' [-fpermissive]
     if (!WITHIN(i, -99, 999)) return i16tostr4sign((int)f);
                                      ~~~~~~~~~~~~~^~~~~~~~
*** [.pio\build\LPC1768\src\src\libs\numtostr.cpp.o] Error 1
```

**Comment by LinFor** [comment]
Here is fixup: https://github.com/LinFor/Marlin/commit/a7326493d57a210fa4a870ada0b60c5584645199

**Comment by blueeagle69** [comment]
Hi 
Daft question.
Does this excellent addition  support PrusaSlicer's Remaining Tine option or is it a standalone feature?
Either way, fantastic.

**Pull request merged**